Have .dockerignore support Dockerfile/.dockerignore

duglin opened the pull request:
If .dockerignore mentions either then the client will send them to the
daemon but the daemon will erase them after the Dockerfile has been parsed
to simulate them never being sent in the first place.

Closes #8330

Signed-off-by: Doug Davis dug@us.ibm.com

Commit:
Have .dockerignore support Dockerfile/.dockerignore

If .dockerignore mentions either then the client will send them to the
daemon but the daemon will erase them after the Dockerfile has been parsed
to simulate them never being sent in the first place.

an events test kept failing for me so I tried to fix that too

Closes #8330

Signed-off-by: Doug Davis <dug@us.ibm.com>

jfrazelle commented on builder/evaluator.go:
```diff
@@ -185,6 +189,27 @@ func (b *Builder) Run(context io.Reader) (string, error) {
+
+	excludes, _ := utils.ReadDockerIgnore(path.Join(b.contextPath, ".dockerignore"))
+	if rm, _ := fileutils.Matches(".dockerignore", excludes); rm == true {
+		os.Remove(path.Join(b.contextPath, ".dockerignore"))
```
should we not be returning errors here?

cpuguy83 replied:
fileutils.Matches only returns an error from filepath.Matches, which only has an error if the pattern is bad, but we know the pattern is good.

jfrazelle replied:
i meant an error on the `os.Remove()`

cpuguy83 replied:
Move along, nothing to see here :)

It would be really really really strange to get an error here... so maybe.

duglin replied:
I can add it if you guys want but it would be pretty odd.

crosbymichael commented on utils/utils.go:
```diff
@@ -492,3 +493,32 @@ func StringsContainsNoCase(slice []string, s string) bool {
+// as use GO's "clean" func to get the shortest/cleanest path for each.
+func ReadDockerIgnore(path string) ([]string, error) {
+	// Note that a missing .dockerignore file isn't treated as an error
+	reader, err := os.Open(path)
```
You are missing a defer here closing this file

duglin replied:
Fixed

crosbymichael commented on builder/evaluator.go:
```diff
@@ -173,6 +179,39 @@ func (b *Builder) Run(context io.Reader) (string, error) {
 	return b.image, nil
 }
 
+func (b *Builder) cleanupDockerignore() {
```
I'm not a fan of functions like this because it's not clear when and where it should be called as it accepts no input and produces no output.  Maybe this should be part of `readContext` and that should return the Dockerfile ast?

Sorry, I don't mean to lag more on this, we are just dealing with methods like this in another area of the codebase so it's something that I noticed.

duglin commented:
@crosbymichael I refactored the reading of the dockerfile into its own function and moved the cleanup stuff into there so it shouldn't look so odd to you any more.  Have a look and let me know.

crosbymichael commented:
@duglin awesome, thanks

crosbymichael commented:
LGTM
